README.md

[jetty]: http://eclipse.org/jetty/
[maven]: http://maven.apache.org/
[java]: https://www.oracle.com/java/index.html
[git]: https://git-scm.com/

[make]: https://www.gnu.org/software/make

NDEx Search REST Service
========================

[![Build Status](https://app.travis-ci.com/cytoscape/ndexsearch-rest.svg?branch=master)](https://app.travis-ci.com/cytoscape/ndexsearch-rest)

Provides integrated search REST service using NDEx as a backend.
This service runs using an embedded [Jetty][jetty] server and is invoked
from the command line. 

Requirements
============

* Centos 7+, Ubuntu 12+, and most other linux distributions should work
* [Java][java] 11+ **(jdk to build)**
* [Make][make] **(to build)**
* [Maven][maven] 3.6 or higher **(to build)**

Dependencies deployed on [NRNB Nexus](https://nrnb-nexus.ucsd.edu/)

* [ndex-enrichment-rest](https://github.com/cytoscape/ndex-enrichment-rest)
* [ndex-java-client](https://github.com/ndexbio/ndex-java-client)
* [ndex-interactome-search](https://github.com/cytoscape/ndex-interactome-search)

Building NDEx Search REST Service  
=================================


Commands below build NDEx Search REST Service assuming machine has [Git][git] command line tools 
installed and above Java modules have been installed:

```Bash
# In lieu of git one can just download repo and unzip it
git clone https://github.com/cytoscape/ndexsearch-rest.git

cd ndexsearch-rest
mvn clean test install
```

The above command will create a jar file under **target/** named  
**ndexsearch-rest-\<VERSION\>-jar-with-dependencies.jar** that
is a command line application

Running NDEx Search REST Service
===============================

The following steps cover how to configure and run integrated search REST service.
In the steps below **ndexsearch.jar** refers to the jar
created previously named **ndexsearch-rest-\<VERSION\>-jar-with-dependencies.jar**

### Step 1 Create directories and configuration file

```bash
# create directory
mkdir -p searchdb/logs searchdb/tasks
cd searchdb

# Generate template configuration file
java -jar ndexsearch.jar --mode exampleconf > ndexsearch.conf
```

The `ndexsearch.conf` file will look like the following:

```bash
# Example configuration file for Search service


# Sets Search database directory
search.database.dir = /tmp

# Sets imageURL to use for results that lack imageURL
search.unset.image.url = http://ndexbio.org/images/new_landing_page_logo.06974471.png

# Sets Search task directory where results from queries are stored
search.task.dir = /tmp/tasks

# Sets HOST URL prefix (value is prefixed to Location header when query is invoked. Can be commented out)
# search.host.url = http://ndexbio.org

# Run Service under embedded Jetty command line parameters
runserver.contextpath = /
runserver.log.dir = /tmp/log
runserver.port = 8080
# Valid log levels DEBUG INFO WARN ERROR ALL
runserver.log.level = INFO

# Sets name of json file containing source results.
# This file expected to reside in search.database.dir directory
source.configurations = source.configurations.json
ndex.user = bob
ndex.password = somepassword
ndex.server = public.ndexbio.org
ndex.useragent = NDExSearch/0.8.0

```

Replace **/tmp** paths with full path location to **searchdb** directory created
earlier.

### Step 2 Create source.configurations.json file

This file (which resides in **search.database.dir**) contains
information about services this integrated search and utilize.

Run the following to create an example **source.configurations.json** file:

```bash
java -jar ndexsearch.jar --mode examplesourceconfig > source.configurations.json
```

The **source.configurations.json** file will look like this:

```bash
{
  "sources" : [ {
    "name" : "enrichment",
    "description" : "This is a description of enrichment source",
    "endPoint" : "http://localhost:8095/enrichment/v1/",
    "uuid" : null
  }, {
    "name" : "interactome-ppi",
    "description" : "This is a description of interactome-ppi service",
    "endPoint" : "http://localhost:8096/interactome/ppi/v1/",
    "uuid" : null
  }, {
    "name" : "interactome-association",
    "description" : "This is a description of interactome-association service",
    "endPoint" : "http://localhost:8096/interactome/geneassociation/v1/",
    "uuid" : null
  } ]
}
```

Each service under **sources** has various fields needed to access that service.

The **uuid** should be set to a unique value for each service.

### Step 4 Run the service

```bash
jav -jar ndexsearch.jar --mode runserver --conf ndexsearch.conf
```



COPYRIGHT AND LICENSE
=====================

TODO

Acknowledgements
================

TODO


HISTORY.md
History
========

0.8.0 (2022-03-11)
-------------

* To be consistent with other IQuery services, version bumped up to `0.8.0`

* Bumped jetty version to `9.4.45.v20220203`

* Bumped jackson-databind to `2.9.10.8`

* Added support for mutation frequency if caller includes `geneAnnotationServices` in query with
  URLs to service endpoints that provide mutation frequency. This includes addition of a mock mutation
  frequency endpoint

* When submitting query the resulting json now includes a `webURL` link that displays the
  results of that query in IQuery

* Fixed possible race conditions during processing of queries by adding `synchronized` wrappers

* Added logging of query genes and input sources. Also added logging of requests ip address and endpoint visited

0.3.6 (2021-02-22)
------------------

* Fixed a bug where if a query had no results the source would be omitted from
  the returned JSON which did not adhere to the expected behavior of web UI. UD-1591

0.3.5 (2020-11-24)
------------------

* Fixed a bug where if `/status` endpoint raised an exception (unlikely to happen), it
  would return a 500 code, but not output json describing the error.

* Bumped jetty version to `9.4.34.v20201102` and jackson to `2.9.10` databind `2.9.10.6`

* Fixed bug on getting results with source filter. If one passed a list of sources delimited
  by a comma with no spaces between the terms, none of the sources would be returned.

* Fixed bug where periodic update of sources would clear sources if one of them 
  (in this case, enrichment) failed when update called. UD-1569.



src/main/java/org/ndexbio/ndexsearch/App.java
package org.ndexbio.ndexsearch;


import ch.qos.logback.classic.Level;
import ch.qos.logback.classic.LoggerContext;
import ch.qos.logback.classic.encoder.PatternLayoutEncoder;
import ch.qos.logback.core.OutputStreamAppender;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.util.Properties;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.PrintStream;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import joptsimple.OptionException;
import joptsimple.OptionParser;
import joptsimple.OptionSet;
import org.eclipse.jetty.server.Handler;
import org.eclipse.jetty.util.RolloverFileOutputStream;
import org.eclipse.jetty.util.log.Log;
import org.eclipse.jetty.server.Server;
import org.eclipse.jetty.server.handler.ContextHandlerCollection;
import org.eclipse.jetty.servlet.ServletContextHandler;
import org.eclipse.jetty.servlet.ServletHolder;
import org.ndexbio.ndexsearch.rest.CorsFilter;
import org.ndexbio.ndexsearch.rest.RequestLoggingFilter;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import org.ndexbio.ndexsearch.rest.model.SourceConfiguration;
import org.ndexbio.ndexsearch.rest.model.SourceConfigurations;
import org.ndexbio.ndexsearch.rest.services.Configuration;
import org.ndexbio.ndexsearch.rest.services.SearchHttpServletDispatcher;

/**
 *
 * @author churas
 */
public class App {
    
    static Logger _logger = LoggerFactory.getLogger(App.class);

    
    /**
     * Sets logging level valid values DEBUG INFO WARN ALL ERROR
     */
    public static final String RUNSERVER_LOGLEVEL = "runserver.log.level";
    /**
     * Sets log directory for embedded Jetty
     */
    public static final String RUNSERVER_LOGDIR = "runserver.log.dir";
    
    /**
     * Sets port for embedded Jetty
     */
    public static final String RUNSERVER_PORT = "runserver.port";
        
    /**
     * Sets context path for embedded Jetty
     */
    public static final String RUNSERVER_CONTEXTPATH = "runserver.contextpath";
    
    
    public static final String MODE = "mode";
    public static final String CONF = "conf";    
    public static final String EXAMPLE_CONF_MODE = "exampleconf";
    public static final String EXAMPLE_SOURCE_CONF_MODE = "examplesourceconfig";
    public static final String RUNSERVER_MODE = "runserver";
    public static final String TESTQUERY_MODE = "testquery";
    
    public static final String SUPPORTED_MODES = ", " + EXAMPLE_CONF_MODE +
    						    ", " + EXAMPLE_SOURCE_CONF_MODE +
                                                    ", " + RUNSERVER_MODE + 
                                                    ", " + TESTQUERY_MODE;
    
    public static void main(String[] args){

        final List<String> helpArgs = Arrays.asList("h", "help", "?");
        try {
            OptionParser parser = new OptionParser() {

                {
                    accepts(MODE, "Mode to run. Supported modes: " + SUPPORTED_MODES).withRequiredArg().ofType(String.class).required();
                    accepts(CONF, "Configuration file")
                            .withRequiredArg().ofType(String.class);
                    acceptsAll(helpArgs, "Show Help").forHelp();
                }
            };
            
            OptionSet optionSet = null;
            try {
                optionSet = parser.parse(args);
            } catch (OptionException oe) {
                System.err.println("\nThere was an error parsing arguments: "
                        + oe.getMessage() + "\n\n");
                parser.printHelpOn(System.err);
                System.exit(1);
            }

            //help check
            for (String helpArgName : helpArgs) {
                if (optionSet.has(helpArgName)) {
                    System.out.println("\n\nHelp\n\n");
                    parser.printHelpOn(System.out);
                    System.exit(2);
                }
            }
            
            String mode = optionSet.valueOf(MODE).toString();

            if (mode.equals(EXAMPLE_CONF_MODE)){
                System.out.println(generateExampleConfiguration());
                System.out.flush();
                return;
            }
            if (mode.equals(EXAMPLE_SOURCE_CONF_MODE)){
                System.out.println(generateExampleSourceConfigurations());
                System.out.flush();
                return;
            }
                        
            if (mode.equals(RUNSERVER_MODE)){
                Configuration.setAlternateConfigurationFile(optionSet.valueOf(CONF).toString());
                Properties props = getPropertiesFromConf(optionSet.valueOf(CONF).toString());
                ch.qos.logback.classic.Logger rootLog = 
        		(ch.qos.logback.classic.Logger) org.slf4j.LoggerFactory.getLogger(Logger.ROOT_LOGGER_NAME);
                rootLog.setLevel(Level.toLevel(props.getProperty(App.RUNSERVER_LOGLEVEL, "INFO")));
                String logDir = props.getProperty(App.RUNSERVER_LOGDIR, ".");
                RolloverFileOutputStream os = new RolloverFileOutputStream(logDir + File.separator + "ndexsearch_yyyy_mm_dd.log", true);
			  
			
				LoggerContext lc = (LoggerContext) LoggerFactory.getILoggerFactory();

				PatternLayoutEncoder logEncoder = new PatternLayoutEncoder();
				logEncoder.setContext(lc);
				logEncoder.setPattern("[%date]\t%msg%n");
				logEncoder.start();
               
				
                RolloverFileOutputStream requestOS = new RolloverFileOutputStream(logDir + File.separator + "requests_yyyy_mm_dd.log", true);
				
				OutputStreamAppender osa = new OutputStreamAppender();
				osa.setOutputStream(requestOS);
				osa.setContext(lc);
				osa.setEncoder(logEncoder);
				osa.setName(RequestLoggingFilter.REQUEST_LOGGER_NAME + "appender");
				osa.start();
				ch.qos.logback.classic.Logger requestLog = 
					  (ch.qos.logback.classic.Logger) lc.getLogger(RequestLoggingFilter.REQUEST_LOGGER_NAME);
                requestLog.setLevel(Level.toLevel("INFO"));
				requestLog.setAdditive(false);
				requestLog.addAppender(osa);
		
                final int port = Integer.valueOf(props.getProperty(App.RUNSERVER_PORT, "8080"));
                System.out.println("\nSpinning up server for status invoke: http://localhost:" + Integer.toString(port) + Configuration.V_ONE_PATH + "/status\n\n");
                System.out.flush();
                
                //We are creating a print stream based on our RolloverFileOutputStream
		PrintStream logStream = new PrintStream(os);

                //We are redirecting system out and system error to our print stream.
		System.setOut(logStream);
		System.setErr(logStream);

                final Server server = new Server(port);

                final ServletContextHandler webappContext = new ServletContextHandler(server, props.getProperty(App.RUNSERVER_CONTEXTPATH, "/"));
                
                HashMap<String, String> initMap = new HashMap<>();
                initMap.put("resteasy.servlet.mapping.prefix", "/");
                initMap.put("javax.ws.rs.Application", "org.ndexbio.ndexsearch.rest.SearchApplication");
                final ServletHolder restEasyServlet = new ServletHolder(
                     new SearchHttpServletDispatcher());
                
                restEasyServlet.setInitOrder(1);
                restEasyServlet.setInitParameters(initMap);
                webappContext.addServlet(restEasyServlet, "/*");
                webappContext.addFilter(CorsFilter.class, "/*", null);
				webappContext.addFilter(RequestLoggingFilter.class, "/*", null);
                ContextHandlerCollection contexts = new ContextHandlerCollection();
                contexts.setHandlers(new Handler[] { webappContext });
 
                server.setHandler(contexts);
                
                server.start();
                Log.getRootLogger().info("Embedded Jetty logging started.", new Object[]{});
	    
                System.out.println("Server started on port " + port);
                server.join();
                return;
            }
            
        }
        catch(Exception ex){
            ex.printStackTrace();
        }
    }
    
    public static Properties getPropertiesFromConf(final String path) throws IOException, FileNotFoundException {
        Properties props = new Properties();
        props.load(new FileInputStream(path));
        return props;
    }
    
    
    /**
     * Generates an example source configurations file
     * file as String
     * @return String 
     * @throws Exception If there was a problem generating output
     */
    public static String generateExampleSourceConfigurations() throws Exception {
        SourceConfiguration scA = new SourceConfiguration();
        scA.setDescription("This is a description of enrichment source");
        scA.setName("enrichment");
        scA.setEndPoint("http://localhost:8095/enrichment/v1/");
     
        
        SourceConfiguration scB = new SourceConfiguration();
        scB.setDescription("This is a description of interactome-ppi service");
        scB.setName("interactome-ppi");
        scB.setEndPoint("http://localhost:8096/interactome/ppi/v1/");
   
        
        SourceConfiguration scC = new SourceConfiguration();
        scC.setDescription("This is a description of interactome-association service");
        scC.setName("interactome-association");
        scC.setEndPoint("http://localhost:8096/interactome/geneassociation/v1/");  
        
        SourceConfigurations scs = new  SourceConfigurations();
        scs.setSources(Arrays.asList(scA, scB, scC));
        ObjectMapper mappy = new ObjectMapper();
        
        return mappy.writerWithDefaultPrettyPrinter().writeValueAsString(scs);
    }

    /**
     * Generates example Configuration file writing to standard out
     * @return Example configuration as String
     */
    public static String generateExampleConfiguration() {
        StringBuilder sb = new StringBuilder();
        sb.append("# Example configuration file for Search service\n\n");
        
        sb.append("\n# Sets Search database directory\n");
        sb.append(Configuration.DATABASE_DIR + " = /tmp\n");
        
        sb.append("\n# Sets imageURL to use for results that lack imageURL\n");
        sb.append(Configuration.UNSET_IMAGE_URL);
        sb.append(" = http://ndexbio.org/images/new_landing_page_logo.06974471.png\n");
        
        sb.append("\n# Sets Search task directory where results from queries are stored\n");
        sb.append(Configuration.TASK_DIR + " = /tmp/tasks\n");
        
        sb.append("# Sets HOST URL prefix (value is prefixed to Location header when query is invoked. Can be commented out)\n");
        sb.append("# " + Configuration.HOST_URL + " = http://ndexbio.org\n");
		
		sb.append("\n# Sets Web App URL prefix that is returned from a query (if unset, null is returned in Task)\n");
		sb.append("# " + Configuration.WEB_URL + " = https://ndexbio.org/iquery\n");
        
        sb.append("\n# Run Service under embedded Jetty command line parameters\n");
        sb.append(App.RUNSERVER_CONTEXTPATH + " = /\n");
        sb.append(App.RUNSERVER_LOGDIR + " = /tmp/log\n");
        sb.append(App.RUNSERVER_PORT + " = 8080\n");
        sb.append("# Valid log levels DEBUG INFO WARN ERROR ALL\n");
        sb.append(App.RUNSERVER_LOGLEVEL + " = INFO\n");
        
        sb.append("\n# Sets name of json file containing source results.\n# This file ");
        sb.append("expected to reside in " + Configuration.DATABASE_DIR + " directory\n");
        sb.append(Configuration.SOURCE_CONFIGURATIONS_JSON_FILE+ " = " + Configuration.SOURCE_CONFIGURATIONS_JSON_FILE + ".json\n");
        sb.append(Configuration.NDEX_USER+ " = bob\n");
        sb.append(Configuration.NDEX_PASS+ " = somepassword\n");
        sb.append(Configuration.NDEX_SERVER+ " = public.ndexbio.org\n");
        sb.append(Configuration.NDEX_USERAGENT+ " = NDExSearch/1.0\n");
        return sb.toString();
    }    
}
